### ISSUE TITLE
[FEATURE] Alternate Measurement Systems

### ISSUE THREAD

**Issue by lenaleciel** [Eranziel/foundryvtt-lancer#396]
**Is your feature request related to a problem? Please describe.**
Currently, the only measurement system available for the ruler in the Lancer module is one where each diagonal space is treated as one space of movement. 

**Describe the solution you'd like**
Different tables may use different measurement systems for Lancer as it is not explicitly written for grids or hexes, and thus support for different measurement systems would be appreciated. For example, some alternate systems that people use include taxicab (in which diagonals are measured by the sum of both the horizontal and vertical movement) and alternating diagonals (in which each diagonal alternates between costing 1 and 2 spaces of movement, in a 1-2-1-2 pattern, effectively costing diagonals as 1.5, rounded down to the nearest full number.)

**Describe alternatives you've considered**
Theoretically a module would be enough for this, but such modules do not seem to exist, as the intended behavior seems to already exist in multiple popular game systems, such as DnD5e and PF2e. I'd do it myself if I had any idea as to how Foundry works, though if this task is simple enough, I might end up just looking up things, and submitting a pull request myself.

**Discord contact info**
lcl#4495

**Comment by msprijatelj** [comment]
Would this be covered by a combination of the modules Drag Ruler and Lancer Speed Provider?

**Comment by lenaleciel** [comment]
Unless there's some functionality to Drag Ruler and Lancer Speed Provider that's not in the default module settings when both are installed, no.

**Comment by Eranziel** [comment]
This is a good request, but I'm going to leave it to others to work on. If I recall correctly, Pathfinder and/or D&D5 have 1-diagonal grid measurement, so they may be good examples to look at.

**Comment by ScyllPoesis** [comment]
I played with this during my time tinkering on 5e modules.

The way you do it is by hooking into foundry's `SquareGrid` (https://foundryvtt.com/api/SquareGrid.html) proto and overriding the native measuring with your own function. Relevant chunks from 5e code:
```
Hooks.on("canvasInit", function() {
  // Extend Diagonal Measurement
  canvas.grid.diagonalRule = game.settings.get("dnd5e", "diagonalMovement");
  SquareGrid.prototype.measureDistances = measureDistances;
});
```
```
export const measureDistances = function(segments, options={}) {
  if ( !options.gridSpaces ) return BaseGrid.prototype.measureDistances.call(this, segments, options);

  // Track the total number of diagonals
  let nDiagonal = 0;
  const rule = this.parent.diagonalRule;
  const d = canvas.dimensions;

  // Iterate over measured segments
  return segments.map(s => {
    let r = s.ray;

    // Determine the total distance traveled
    let nx = Math.abs(Math.ceil(r.dx / d.size));
    let ny = Math.abs(Math.ceil(r.dy / d.size));

    // Determine the number of straight and diagonal moves
    let nd = Math.min(nx, ny);
    let ns = Math.abs(ny - nx);
    nDiagonal += nd;

    // Alternative DMG Movement
    if (rule === "5105") {
      let nd10 = Math.floor(nDiagonal / 2) - Math.floor((nDiagonal - nd) / 2);
      let spaces = (nd10 * 2) + (nd - nd10) + ns;
      return spaces * canvas.dimensions.distance;
    }

    // Euclidean Measurement
    else if (rule === "EUCL") {
      return Math.round(Math.hypot(nx, ny) * canvas.scene.data.gridDistance);
    }

    // Standard PHB Movement
    else return (ns + nd) * canvas.scene.data.gridDistance;
  });
};
```

**Comment by Eranziel** [comment]
Implemented by #456.
